extensions/mssql/src/objectExplorerNodeProvider/objectExplorerNodeProvider.ts

```diff
@@ -190,92 +167,105 @@ export class MssqlObjectExplorerNodeProvider extends ProviderBase implements sql
 		}
 	}
 
-	private getSessionForNode(node: TreeNode): Session {
-		let rootNode: DataServicesNode = undefined;
-		while (rootNode === undefined && node !== undefined) {
+	private getSqlClusterSessionForNode(node: TreeNode): SqlClusterSession {
+		let sqlClusterSession: SqlClusterSession = undefined;
+		while (node !== undefined) {
 			if (node instanceof DataServicesNode) {
-				rootNode = node;
+				sqlClusterSession = node.sqlClusterSession;
 				break;
 			} else {
 				node = node.parent;
 			}
 		}
-		if (rootNode) {
-			return rootNode.session;
-		}
-		// Not found
-		return undefined;
+		return sqlClusterSession;
 	}
 
-	async findNodeForContext<T extends TreeNode>(explorerContext: sqlops.ObjectExplorerContext): Promise<T> {
+	async findSqlClusterNodeBySqlContext<T extends TreeNode>(sqlContext: ICommandObjectExplorerContext | sqlops.ObjectExplorerContext): Promise<T> {
 		let node: T = undefined;
-		let session = await this.findSessionForConnection(explorerContext.connectionProfile);
+		let sqlOeContext = 'explorerContext' in sqlContext ? sqlContext.explorerContext : sqlContext;
+		let sqlConnProfile = sqlOeContext.connectionProfile;
+		let session = this.findSqlClusterSessionBySqlConnProfile(sqlConnProfile);
 		if (session) {
-			if (explorerContext.isConnectionNode) {
+			if (sqlOeContext.isConnectionNode) {
 				// Note: ideally fix so we verify T matches RootNode and go from there
-				node = <T><any>session.root;
+				node = <T><any>session.rootNode;
 			} else {
 				// Find the node under the session
-				node = <T><any>await session.root.findNodeByPath(explorerContext.nodeInfo.nodePath, true);
+				node = <T><any>await session.rootNode.findNodeByPath(sqlOeContext.nodeInfo.nodePath, true);
 			}
 		}
 		return node;
 	}
 
-	private async findSessionForConnection(connectionProfile: sqlops.IConnectionProfile): Promise<Session> {
-		for (let session of this.sessionMap.values()) {
-			if (session.connection && await session.connection.isMatch(connectionProfile)) {
-				return session;
-			}
-		}
-		return undefined;
+	public findSqlClusterSessionBySqlConnProfile(sqlConnProfile: sqlops.IConnectionProfile): SqlClusterSession {
+		return Array.from(this.sessionMap.values())
```

Comment:
This does an array copy right? Can we just use the previous logic which avoids this?

Reply:
rolled back.